docs/WinRTAPIContracts.md

```diff
@@ -0,0 +1,76 @@
+# Windows App SDK API Contracts
+
+## Summary
+Customers trust Microsoft that the products we make are sound, and will work.  Windows App SDK (WASDK) has recently released version 1.0.  As WASDK evolvs and more developers use WASDK in their own projects WASDK needs a way to guarantee that the API’s that ship are stable and that API versions are well documented.
+To keep customers’ trust in WASDK, and to make sure the WASDK API’s are
+1.  Binary compatible
+2.  Stable
+3.  Customizable so developers can include only the parts they want
+
+Contracts, and contract versions should be added to WASDK.
+
+## Architecture Decisions
+1. Contract boundaries
+Contracts will be split among the winmd boundaries.  This is a good medium between having 1 contract for everything, and 1 contract per class.
+2. Contract numbering.
+Contracts will be numbered with MAJOR.MINOR.  Any update to the API is a MAJOR change.  Any servicing update is a MINOR change.
+3. Enforcement
+The build systems will have enforcement gates to make sure that contracts are present, up to date, and there are no conflicts
+
+## Example
+    namespace Microsoft.Windows.System
+    {
+        [contractversion(2.0)]
+        apicontract MicrosoftWindowsSystem
+        
+        // version 2.0
+        [Contract(MicrosoftWindowsSystem(2.0)]
+        public enum Scope
+        {
+            user,
+            machine
+        }
+
+        [Contract(MicrosoftWindowsSystem(2.0)]
+        public class EnvironmentManager
+        {
+            public String GetEnvironmentVariable(String name, Scope scope);
+        }
+        
+        // servicing
+        [Contract(MicrosoftWindowsSystem(1.5)]        
+        public class EnvironmentManager
+        {
+            public String GetEnvironmentVariable(String name, String scope);
+        }
+
+        // base version
+        [contract(MicrosoftWindowsSystem, 1.0),]
+        public class EnvironmentManager
+        {
+            public String GetEnvironmentVariable(String name);
+        }
+    }
+
+## 1.1 ship blocker
+Any API's that changed between 1.0 and 1.1 need to add a contract or 1.1 will not ship.
+
+## Telemetry
+Because all API's have a contract version data can be used to see
+ 1. How many people are using each API
+ 2. How many people are using each specific version
+
+This data will allow us to better direct our development efforts to help the most developers.
+
+## Gate check-ins
+Because the contracts are used for API versions there will need to be work done in the build system to make sure
+1. All API's have contracts
+2. Each contract is correct.  This means version and name.
+3. All Contracts and dependencies work with each other.
```

Comment:
What about compatibility? Don't we need reports (detection) of **changes**? maybe even enforcement to reject changes that break compat?

Reply:
Yes. We need to enforce these semantics on contract versioning. Once a contract version has gone into a stable release, any changes to types/APIs in that contract must go into a new version, and compat-breaking changes are forbidden. Specifically, after a release
- Method signatures must not change.
- Struct layouts must not change.
- Interfaces must not add, remove, or reorder methods
- All new properties/methods/events added to a runtimeclass must be put on a new interface, which is on a new contract version.
  - When using MIDL3, the new versioned interface is automatically created by putting the contract version on the added property/method/event.
- Ditto for adding static methods/properties/events to a runtimeclass.
- Enum fields can be added under a new contract version. All other enum changes like reordering, changing, or removing fields, is forbidden.

Reply:
Can a check like this happen at build time?  In both a developers environment and the internal build system?  Also, should this list be added to a "Forbidden changes" section?